src/base/emu-i386/kvmmon.S

```diff
@@ -71,6 +73,12 @@ kvm_mon_hlt:
         addl $0x200,%esp
 #endif
         pop %eax
+        pop %eax
+        or %eax,%eax
+        jz skip_cr3
```

Comment:
Maybe instead add another entry
that will reload cr3 and jump to normal
entry (hlt) ?
kvm.c would use that entry only for
TLB flush.
Thoughts?

Reply:
I've implemented an alternative approach using invlpg, which is more precise than reloading cr3, it does use another entry but needs to return (there could be multiple mprotects between regular kvm runs).

In the end the JIT protection invalidations could be replaced by dirty page logging for KVM, but we'd still need the page invalidations for DPMI fn 0507.

Reply:
> use another entry but needs to return (there could be multiple mprotects between regular kvm runs).

Which is why I think reloading cr3 may
be faster. Even if invlpg is slightly faster,
you get multiple returns, which is horrid.
cr3 reload doesn't need to return.

Reply:
In that case I like my original patch better, as it doesn't need the additional ioctls to change `rip` to point to the `cr3` code, and the runtime overhead of checking for cr3==0 is tiny compared to the KVM_RUN ioctl's overhead.

Of course the cr3 reload could be replaced by multiple invlpg loops without multiple returns but then you need to setup a more complex data structure.

Reply:
How about developing a light-weight
entry points? You can put the indirect
jump after hlt, and then you'll be able
to set up any entry.
Just for consideration, your first patch
is also reasonable.

Reply:
ok I reverted to the first patch. If there are more such cases in the future, sure we'd need something more flexible, but I think with just one the or/jz is ok.

For efficiency it would also be possible to mark the monitor pages as "global", but that could be another PR.